Doctests examples for Linear Algebra library as suggested in #18799

SaschaMann opened the pull request:
Issue JuliaLang/LinearAlgebra.jl#371 
### Added Doctests
- `cholfact(A)`
- `kron(A, B)`
- `pinv(M)`
- `nullspace(M)`
- `Diagonal(A::AbstractMatrix), Diagonal(V::AbstractVector)`
- `norm(v), norm(v, Inf), norm(A, Inf)`
- `dot(a, b)`
- `inv(M)`
- `qr(v)`
- `svd(A), svdvals(A)`
- `SymTridiagonal(dv, ev), Tridiagonal(dl, d, du)`, `Tridiagonal(A)`
### Rewritten examples
- `Bidiagonal()`
- `Symmetric()`
- `Hermitian()`

I have rewritten these existing examples to be easier to understand without having to run the code.

`make docs && make -C doc doctest` runs fine, no failures for `stdlib/linalg`. The failures it shows are identical to the ones on the master branch.

First PR, so please let me know if I did something wrong :)

SaschaMann committed:
Added doctests to linalg package

SaschaMann committed:
Rewrote existing examples to use doctests in linalg package

Sacha0 commented on base/linalg/bidiag.jl:
```diff
@@ -22,11 +22,33 @@ matrix with [`full`](:func:`full`). `ev`'s length must be one less than the leng
+  1  ⋅  ⋅  ⋅
+  7  2  ⋅  ⋅
+  ⋅  8  3  ⋅
+  ⋅  ⋅  9  4
```
Perhaps replace `#e is` with `# ev is` for consistency with the rest of the docstring?

Sacha0 commented on base/linalg/bidiag.jl:
```diff
@@ -42,11 +64,33 @@ matrix with [`full`](:func:`full`). `ev`'s length must be one less than the leng
+ 1  ⋅  ⋅  ⋅
+ 7  2  ⋅  ⋅
+ ⋅  8  3  ⋅
+ ⋅  ⋅  9  4
```
Likewise in this section, perhaps replace `#e is` with `# ev is` for consistency with the rest of the docstring?

Sacha0 commented on base/linalg/dense.jl:
````diff
@@ -211,6 +211,25 @@ end
+ 0  1  0  2
+ 2  0  0  0
+ 0  1  0  0
+```
````
Perhaps
```julia
julia> A = [1 2; 3 4]
2×2 Array{Int64,2}:
 1  2
 3  4

julia> B = [1 1; 1 1]
2×2 Array{Int64,2}:
 1  1
 1  1

julia> kron(A, B)
4×4 Array{Int64,2}:
 1  1  2  2
 1  1  2  2
 3  3  4  4
 3  3  4  4
```
would be easier to grasp?

Sacha0 commented on base/linalg/dense.jl:
````diff
@@ -486,9 +505,22 @@ If `factorize` is called on a Hermitian positive-definite matrix, for instance,
-A = diagm(rand(5)) + diagm(rand(4),1); #A is really bidiagonal
-factorize(A) #factorize will check to see that A is already factorized
+```jldoctest
+julia> A = Bidiagonal(ones(5, 5), true) #A is really bidiagonal
````
Calling the `Bidiagonal` constructor to clarify what's going on is a great idea! The remainder of the example relies on `A` being an `Array` though. Perhaps wrap your `Bidiagonal` constructor call in an `Array` constructor call (`Array(Bidiagonal(ones(5, 5), true))`)?

Calling the `Bidiagonal` constructor makes this example sufficiently clear that the comment may no longer  be necessary?

SaschaMann replied:
Ah, I missed the idea behind the second part of the example. I added the `Array` constructor to it.

Sacha0 commented on base/linalg/dense.jl:
```diff
@@ -211,6 +211,25 @@ end
     kron(A, B)
 
 Kronecker tensor product of two vectors or two matrices.
+
```
Perhaps add an example heading?

Sacha0 commented on base/linalg/cholesky.jl:
```diff
@@ -267,6 +267,33 @@ The default is to use `:U`.
 The triangular Cholesky factor can be obtained from the factorization `F` with: `F[:L]` and `F[:U]`.
 The following functions are available for `Cholesky` objects: `size`, `\\`, `inv`, `det`.
 A `PosDefException` exception is thrown in case the matrix is not positive definite.
+
```
Perhaps add an example heading?

Sacha0 commented on base/linalg/dense.jl:
```diff
@@ -486,9 +505,22 @@ If `factorize` is called on a Hermitian positive-definite matrix, for instance,
 will return a Cholesky factorization.
 
 Example:
```
Perhaps change the `Example:` to a heading?

Sacha0 commented on base/linalg/dense.jl:
````diff
@@ -587,6 +619,23 @@ inverting dense ill-conditioned matrices in a least-squares sense,
 
 For more information, see [^issue8859], [^B96], [^S84], [^KY88].
 
+```jldoctest
````
Perhaps add an example heading?

Sacha0 commented on base/linalg/dense.jl:
```diff
@@ -587,6 +619,23 @@ inverting dense ill-conditioned matrices in a least-squares sense,
 
 For more information, see [^issue8859], [^B96], [^S84], [^KY88].
```
@kshyatt, should the `For more information...` group with the other references after the example, should the other references after the example group with the `For more information...` before the example, or neither? Thoughts? Thanks!

Sacha0 commented on base/linalg/dense.jl:
```diff
@@ -640,6 +689,20 @@ end
     nullspace(M)
 
 Basis for nullspace of `M`.
+
```
Perhaps add an example heading?

Sacha0 commented on base/linalg/diagonal.jl:
```diff
@@ -9,12 +9,38 @@ end
     Diagonal(A::AbstractMatrix)
 
 Constructs a matrix from the diagonal of `A`.
+
```
Perhaps add an example heading?

Sacha0 commented on base/linalg/diagonal.jl:
```diff
@@ -9,12 +9,38 @@ end
     Diagonal(V::AbstractVector)
 
 Constructs a matrix with `V` as its diagonal.
+
```
Perhaps add an example heading?

Sacha0 commented on base/linalg/generic.jl:
```diff
@@ -550,6 +574,21 @@ dot(x::Number, y::Number) = vecdot(x, y)
     ⋅(x,y)
 
 Compute the dot product. For complex vectors, the first vector is conjugated.
+
```
Perhaps add an example heading?

Sacha0 commented on base/linalg/generic.jl:
````diff
@@ -550,6 +574,21 @@ dot(x::Number, y::Number) = vecdot(x, y)
+
+julia> dot(a, b)
+5
+```
````
Perhaps an example of conjugation in the complex case would be useful, e.g. `dot([im, im], [1, 1])`?

Sacha0 reviewed (commented):
@SaschaMann, this is fantastic! A few minor comments on the first half of the PR follow. Thanks and best!

SaschaMann committed:
Improvements & fixes to linalg examples

SaschaMann commented:
@Sacha0 Most of the already existing examples (both in linalg and in other packages) don't have an example heading, so I intentionally didn't add any. Imo the it's clear from the different formatting and context that something is an example unless there's additional text to it. However, I could add them if you prefer that.

Thanks for your comments!

tkelman commented on base/linalg/dense.jl:
```diff
@@ -211,6 +211,25 @@ end
+ 1  2
+ 3  4
+
+julia> B = [1 1; 1 1]
```
might be more interesting if B isn't all ones here

Sacha0 replied:
Apologies, all ones was my suggestion (to make what `kron` does easy to grasp). Can you think of an equally clear but more exciting example? Best!

Sacha0 replied:
Perhaps `[im 1; 1 1]` would be slightly more interesting but similarly easy to grasp?

tkelman replied:
or in a different spot, but yeah at least one entry not equal to 1.0

Sacha0 commented:
> I applied the changes you suggested.

The changes so far look great!

> Most of the already existing examples (both in linalg and in other packages) don't have an example heading, so I intentionally didn't add any. Imo the it's clear from the different formatting and context that something is an example unless there's additional text to it. However, I could add them if you prefer that.

The fifth item in [the documentation documentation](http://docs.julialang.org/en/release-0.5/manual/documentation/)'s list of conventions (about two pages down) suggests segregating examples under an appropriate heading. ([Those conventions were codified only a few months ago](https://github.com/JuliaLang/julia/pull/15136), so much of the existing documentation does not follow them. #8966 contains additional discussion on evolution of documentation conventions; evolving towards more thorough, structured documentation seems to be the plan.)

That said, as-is is fine by me. Cheers if you prefer to omit example headings, and likewise cheers if you prefer to add them. Whichever stokes your enthusiasm! :)

Best!

SaschaMann committed:
Added more examples/doctests to linalg package

SaschaMann commented:
I just added some more examples.

Also decided to add the example headings to all examples in the file. The reasons you gave for the headings were compelling.

Tests still pass fine:

> Document: stdlib/linalg
> -----------------------
> 1 items passed all tests:
>  177 tests in default
> 177 tests in 1 items.
> 177 passed and 0 failed.
> Test passed.

Sacha0 commented on base/linalg/lu.jl:
```diff
@@ -115,6 +115,24 @@ The relationship between `F` and `A` is
+ 4  3
+ 6  3
+
+julia> lufact(A)
```
Duplicate of the equivalent line just below beginning `F = ...`?

SaschaMann replied:
Missed that, when copy pasting from the REPL. Fixed it.

Sacha0 reviewed (commented):
Looking great so far! A few minor comments follow below. Thanks!

SaschaMann committed:
changed example headings & removed duplicate line

SaschaMann commented on base/linalg/cholesky.jl:
```diff
@@ -161,6 +161,25 @@ end
 Compute the Cholesky factorization of a positive definite matrix `A`
 and return the UpperTriangular matrix `U` such that `A = U'U`.
+
+**Example**
```
Yep, you're right. I should have checked the docs instead of just looking at the existing example headings in the file. My bad, fixed it.

SaschaMann commented on base/linalg/cholesky.jl:
```diff
@@ -203,6 +229,19 @@ The same as `cholfact`, but saves space by overwriting the input `A`, instead
+ 2  50
+
+ julia> cholfact!(A)
+ ERROR: InexactError()
```
`cholfact()` has an example with proper usage. the description of the `!`-version just refers to `cholfact()` as well instead of repeting itself, so imo it makes sense to not create duplicate examples, too. The only addition to the description is the mention of the error, which is what this demonstrates.

SaschaMann commented:
Changed the Example headings to `# Example`, removed the duplicate line and added a more interesting example to `kron(A, B)`.

I went through pretty much all the functions in linalg and for the ones that are still without example, I can't come up with anything helpful so I won't add anything after the most recent commit. (Except for fixes or  tweaks to what exists of course, if you suggest anything)

nalimilan commented:
Just wanted to note PRs like that are really cool! Glad we have a Sa(s)cha team writing and reviewing this... :-)

Sacha0 reviewed (approved):
@SaschaMann, this is great work! Much thanks!

andreasnoack commented:
@SaschaMann There seems to be a small conflict. Can you please rebase? When that is done I think it can be merged.

SaschaMann commented:
Hope this worked properly.

I just saw that in the first commit 5b90333 I added the `linalg.rst` as well but afterwards I didn't. To make sure that this doesn't cause issues: Does the `linalg.rst` have to be part of the PR or is it built automatically?

Sacha0 commented:
> Does the linalg.rst have to be part of the PR or is it built automatically?

The PR should include all .rst files that `make docs` modifies. That said, missed .rst files usually get incorporated into other PRs rapidly, so do not worry overmuch about missed .rst files :).

(Two minor notes for the future[1]:

1. Suppressing noncritical output via semicolon termination helps keep examples compact. For example, in the `Bidiagonal` constructor examples, semicolon-terminating the `dv` and `ev` definitions (e.g. `dv = [1; 2; 3; 4];`) would reduce the example's length by almost a factor of two.

2. Writing vector literals with commas seems more idiomatic than using semicolons. For example, `dv = [1, 2, 3, 4]` produces a `Vector{Int64}` (identical to the result of the more general matrix literal `dv = [1; 2; 3; 4]`).

[1] No need to integrate these suggestions in this already great PR! I would advocate merging this PR as is. Of course if you are enthusiastic about integrating either of these suggestions, please enjoy! I would suggest doing so in followup PRs, smaller PRs being easier to review and merge rapidly.)

Thanks again and best!

SaschaMann commented:
@Sacha0 Thanks for the feedback! I might integrate your suggestions later
on if I find some spare time between classes or so.

And thanks again to everyone who helped throughout this PR and the original
issue, I learned quite a lot from it :)

Sacha0 commented:
Barring objections or requests for time, I plan to merge this Wednesday (PDT). Best!

tkelman commented on base/linalg/symmetric.jl:
````diff
@@ -32,13 +54,35 @@ end
+# Example
+
+```jldoctest
+julia> A = [1 0 2 0 3; 0 4 0 5 0; 6 0 7 0 8; 0 9 0 1 0; 2 0 3 0 4]
````
probably better to use an example that's complex to make the distinction clearer?

SaschaMann committed:
Better example for Hermitian()

SaschaMann committed:
added missing .rst file

SaschaMann commented:
Improved the example for `Hermitian()` as suggested and added the missing .rst file.

Sacha0 commented:
Thanks for the great work @SaschaMann!
